core/trie/proof_test.go

```diff
@@ -1,1139 +1,827 @@
-		rightProof, err := trie.GetProof(proofKeys[1], tri)
-		require.NoError(t, err)
-		proofs := [2][]trie.ProofNode{nil, rightProof}
-		rootCommitment, err := tri.Root()
-		require.NoError(t, err)
-		verif, err := trie.VerifyRangeProof(rootCommitment, keys, values, proofKeys, proofValues, proofs, crypto.Pedersen)
-		require.NoError(t, err)
-		require.True(t, verif)
-	})
+func TestGappedRangeProof(t *testing.T) {
+	t.Parallel()
+	t.Skip("gapped keys will sometimes succeed, the current proof format is not able to handle this")
 
-	t.Run("VPR left proof, all inner keys, right proof with non-set key", func(t *testing.T) {
-		zeroFeltBytes := new(felt.Felt).SetUint64(0).Bytes()
-		zeroLeafkey := trie.NewKey(251, zeroFeltBytes[:])
+	tr, records := nonRandomTrie(t, 5)
+	root, err := tr.Root()
+	require.NoError(t, err)
 
-		threeFeltBytes := new(felt.Felt).SetUint64(3).Bytes()
-		threeLeafkey := trie.NewKey(251, threeFeltBytes[:])
+	first, last := 1, 4
+	proof := trie.NewProofNodeSet()
+	err = tr.GetRangeProof(records[first].key, records[last].key, proof)
+	require.NoError(t, err)
 
-		tri := build4KeyTrie(t)
-		keys := []*felt.Felt{new(felt.Felt).SetUint64(1), new(felt.Felt).SetUint64(2)}
-		values := []*felt.Felt{new(felt.Felt).SetUint64(5), new(felt.Felt).SetUint64(6)}
-		proofKeys := [2]*trie.Key{&zeroLeafkey, &threeLeafkey}
-		proofValues := [2]*felt.Felt{new(felt.Felt).SetUint64(4), nil}
-		leftProof, err := trie.GetProof(proofKeys[0], tri)
-		require.NoError(t, err)
-		rightProof, err := trie.GetProof(proofKeys[1], tri)
-		require.NoError(t, err)
+	keys := []*felt.Felt{}
+	values := []*felt.Felt{}
+	for i := first; i <= last; i++ {
+		if i == (first+last)/2 {
+			continue
+		}
 
-		proofs := [2][]trie.ProofNode{leftProof, rightProof}
-		rootCommitment, err := tri.Root()
-		require.NoError(t, err)
+		keys = append(keys, records[i].key)
+		values = append(values, records[i].value)
+	}
 
-		verif, err := trie.VerifyRangeProof(rootCommitment, keys, values, proofKeys, proofValues, proofs, crypto.Pedersen)
-		require.NoError(t, err)
-		require.True(t, verif)
-	})
+	_, err = trie.VerifyRangeProof(root, records[first].key, keys, values, proof)
+	require.Error(t, err)
 }
 
-func TestMergeProofPaths(t *testing.T) {
-	t.Run("3Key Trie no duplicates and all values exist in merged path", func(t *testing.T) {
-		tri := build3KeyTrie(t)
-
-		zeroFeltBytes := new(felt.Felt).SetUint64(0).Bytes()
-		zeroLeafkey := trie.NewKey(251, zeroFeltBytes[:])
+func TestEmptyRangeProof(t *testing.T) {
+	t.Parallel()
 
-		twoFeltBytes := new(felt.Felt).SetUint64(2).Bytes()
-		twoLeafkey := trie.NewKey(251, twoFeltBytes[:])
-
-		proofKeys := [2]*trie.Key{&zeroLeafkey, &twoLeafkey}
+	tr, records := randomTrie(t, 1000)
+	root, err := tr.Root()
+	require.NoError(t, err)
 
-		proofs, err := trie.GetBoundaryProofs(proofKeys[0], proofKeys[1], tri)
-		require.NoError(t, err)
+	cases := []struct {
+		pos int
+		err bool
+	}{
+		{len(records) - 1, false},
+		{500, true},
```

Comment:
I guess this is testing the special case:
`// Special case: there is a provided proof but no key-value pairs, make sure regenerated trie has no more values`
Why is the proof invalid if there are additional elements? Would this not be equivalent to proving non-inclusion?

Reply:
You're right, I'm not sure why geth restricts that there should be no more values given a zero element proof. I'm keen on removing the `hasRightElement` condition for this case, but before that let me try to ask them on this.

Reply:
I got the answer. It's a snap sync specification detail.

Based on Ethereum's snap sync [specs](https://github.com/ethereum/devp2p/blob/master/caps/snap.md#getaccountrange-0x00), it says:

> If no accounts exist between startingHash and limitHash, then the first (if any) account after limitHash must be provided.

So it's guaranteed that empty-range proof can only exist if no more elements are on the right side. Else it's a invalid response.

I would just keep the implementation as it is for now with the added comments. Once we formalize the specification we can decide if we want to keep it or change it.